File: QID-757-MVAcarrankings/MVAcarrankings.py
from sklearn.isotonic import IsotonicRegression
import matplotlib.pyplot as plt

# the reported preference orderings
x = list(range(1, 7))
# the estimated preference orderings according to the additive model (16.1) and 
# the metric solution (Table 16.6) in MVA
y = [0.84, 2.84, 3.16, 3.34, 5.66, 5.16]

gp = IsotonicRegression()
y_gp = gp.fit_transform(x, y)

fig, ax = plt.subplots(figsize = (7, 7))
ax.plot(x, y_gp, c = "k")
ax.scatter(x, y, c = "r")
for i in range(0, len(y)):
    ax.text(x[i]-0.05, y_gp[i]+0.1, "car" + str(i+1), fontsize = 14)
plt.xlabel("revealed rankings", fontsize = 14)
plt.ylabel("estimated rankings", fontsize = 14)
plt.title("Car rankings", fontsize = 16)
plt.show()

File: QID-757-MVAcarrankings/MVAcarrankings.m
%% clear all variables and console and close windows
clear
clc
close all

%% the reported preference orderings
x = (1:6)';

% the estimated preference orderings according to the additive model (16.1) 
% and the metric solution (Table 16.6) in book
y = [0.84, 2.84, 3.16, 3.34, 5.66, 5.16]';
z = [x, y];

%% use PAV algorithm
gp = pava(y);

%% the reported preference orderings
gp = [x, gp];

%% plot
figure(1)
hold on
plot(gp, z, 'w')
xlabel('Revealed rankings', 'FontSize', 12)
ylabel('Estimated rankings', 'FontSize', 12)
title('Car rankings', 'FontSize', 14)
box on
line(gp(:, 1), gp(:, 2), 'LineWidth', 2)
scatter(x, y, 'MarkerEdgeColor', 'red', 'MarkerFaceColor', 'red')%'filled', true)
xlim([0.5, 6.5])
labels = {'car1', 'car2', 'car3', 'car4', 'car5', 'car6'};
text(x, y - 0.1, labels, 'FontSize',12)
hold off
